staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -208,19 +211,12 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 			if r.WatchListPageSize != 0 {
 				pager.PageSize = r.WatchListPageSize
 			}
-			// Pager falls back to full list if paginated list calls fail due to an "Expired" error on the 2nd page or later,
-			// but still my return an "Expired" error if the 1st page fails with "Expired" or the full list fails with "Expired".
+			// Pager falls back to full list if paginated list calls fail due to an "Expired" error on continuation pages.
 			list, err = pager.List(context.Background(), options)
-			if apierrs.IsResourceExpired(err) {
-				// For Kubernetes 1.16 and earlier, if the watch cache is disabled for a resource, list requests
-				// with LastSyncResourceVersion set to a non-zero ResourceVersion will fail if the exact ResourceVersion
-				// requested is expired (e.g. an etcd compaction has remove it).
-				// To prevent the reflector from getting stuck retrying a list for an expired resource version in this
-				// case, we set ResourceVersion="" and list again to re-establish reflector to the latest available
-				// ResourceVersion, using a consistent read from etcd. This is also safe to do if watch cache is enabled
-				// and the list request returned a "Expired" error.
-				options = metav1.ListOptions{ResourceVersion: ""}
-				list, err = pager.List(context.Background(), options)
+			if apierrs.IsResourceExpired(err) || apierrs.IsGone(err) {
```

Comment:
Reviewers: Worth introducing a new apierrors function to check for all error types that have a  HTTP 410 (Gone) status code? `IsResourceExpired` and `IsGone` check status reason.